provider/cloudflare/cloudflare.go

```diff
@@ -516,6 +617,32 @@ func (p *CloudFlareProvider) listDNSRecordsWithAutoPagination(ctx context.Contex
 	return records, nil
 }
 
+// listCustomHostnamesWithPagination performs automatic pagination of results on requests to cloudflare.CustomHostnames
+func (p *CloudFlareProvider) listCustomHostnamesWithPagination(ctx context.Context, zoneID string) ([]cloudflare.CustomHostname, error) {
+	var chs []cloudflare.CustomHostname
+	page := 1
+	for {
+		pageCustomHostnameListResponse, resultInfo, err := p.Client.CustomHostnames(ctx, zoneID, page, cloudflare.CustomHostname{})
+		if err != nil {
+			var apiErr *cloudflare.Error
+			if errors.As(err, &apiErr) {
+				if apiErr.ClientRateLimited() || apiErr.StatusCode >= http.StatusInternalServerError {
+					// Handle rate limit error as a soft error
+					return nil, provider.NewSoftError(err)
+				}
+			}
+			return nil, err
+		}
+
+		chs = append(chs, pageCustomHostnameListResponse...)
+		if resultInfo.TotalPages >= page {
+			break
+		}
+		page++
```

Comment:
Cloudflare's SDK is annoyingly non-contentant, ie. `CustomHostnames()` vs  `ListDNSRecordsParams()` so I'm not sure if it would work but could you use the ```		if params.ResultInfo.Done() {
``` check similar to the `listDNSRecordsWithAutoPagination` function above instead of ```if resultInfo.TotalPages >= page```
?

Reply:
Right, I also didn't like implementing it like that at all, but the cloudflare-go v0 SDK `CustomHostnames()` pagination support is very basic. In contrast with the `ListDNSRecords()` it doesn't have `params` (`params.resultInfo`) parameters - `per_page` is hardcoded to 50, and there is only a `page` parameter.
However, upon checking the [pagination methods](https://github.com/cloudflare/cloudflare-go/blob/v0/pagination.go#L16-L48), it seems we could use the [ResultInfo](https://github.com/cloudflare/cloudflare-go/blob/v0/cloudflare.go#L448-L457) for the only `page` parameter.

It definitely makes the external-dns cloudflare code nicer, thank you for the suggestion!
The code is updated, please have another look.

Reply:
Looks great!